MatrixSDK/Background/MXBackgroundPushRulesManager.swift

```diff
@@ -0,0 +1,201 @@
+// 
+// Copyright 2020 The Matrix.org Foundation C.I.C
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+
+/// Background push rules manager. Does work independent from a `MXNotificationCenter`.
+@objcMembers public class MXBackgroundPushRulesManager: NSObject {
+    
+    private let restClient: MXRestClient
+    private var pushRulesResponse: MXPushRulesResponse? {
+        didSet {
+            var tmpRules: [MXPushRule] = []
+            
+            if let global = pushRulesResponse?.global {
+                if let rules = global.override as? [MXPushRule] {
+                    tmpRules.append(contentsOf: rules)
+                }
+                if let rules = global.content as? [MXPushRule] {
+                    tmpRules.append(contentsOf: rules)
+                }
+                if let rules = global.room as? [MXPushRule] {
+                    tmpRules.append(contentsOf: rules)
+                }
+                if let rules = global.sender as? [MXPushRule] {
+                    tmpRules.append(contentsOf: rules)
+                }
+                if let rules = global.underride as? [MXPushRule] {
+                    tmpRules.append(contentsOf: rules)
+                }
+            }
+            
+            flatRules = tmpRules
+        }
+    }
+    private var flatRules: [MXPushRule] = []
+    private var eventMatchConditionChecker: MXPushRuleEventMatchConditionChecker
+    private var memberCountConditionChecker: MXPushRuleRoomMemberCountConditionChecker
+    private var permissionConditionChecker: MXPushRuleSenderNotificationPermissionConditionChecker
+    
+    /// Initializer.
+    /// - Parameter restClient: Rest client to fetch initial push rules.
+    public init(withRestClient restClient: MXRestClient) {
+        self.restClient = restClient
+        eventMatchConditionChecker = MXPushRuleEventMatchConditionChecker()
+        memberCountConditionChecker = MXPushRuleRoomMemberCountConditionChecker(matrixSession: nil)
+        permissionConditionChecker = MXPushRuleSenderNotificationPermissionConditionChecker(matrixSession: nil)
+        super.init()
+        restClient.pushRules { (response) in
```

Comment:
We should avoid to make parallel requests to avoid race condition.

Plus, it looks to me we already have the information. It is in, by order of truth:
- SyncResponseStore
- MXBackgroundSyncMemoryStore (by using MXFileStore)